### DISCUSSION TITLE
Could not retrieve data from stored procedure

### DISCUSSION THREAD

**Discussion by kurebgd** [godror/godror#146] [Q&A] [answered]
Hi, 

I'm relative novice when it comes to working with databases from Go.
I am trying to retrieve data from database via stored procedure.
E.g. Stored procedure is in the form :
`CREATE OR REPLACE PROCEDURE PACKAGE.StoredProcAdd (pnA1 IN NUMBER, pnA2 IN NUMBER, pnOutput OUT NUMBER) AS
BEGIN
 pnOutput := pnA1 + pnA2;
END;
/`

My Go code looks like this:
`package main

import (
	"context"
	"database/sql"
	"database/sql/driver"
	"fmt"
	"io"
	"time"`
`

	_ "github.com/godror/godror"
)

var (
	ctx context.Context
	db  *sql.DB
)

func main() {
	dbalias := "DBALIAS"
	dbloc := fmt.Sprintf("username/password@%s", dbalias)

	db, err := sql.Open("godror", dbloc)
	if err != nil {
		fmt.Printf("Creating db connection: %v\n", err)
	}
	defer db.Close()

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	var outNo driver.Rows

	var pn1, pn2 int
	pn1 = 5
	pn2 = 7

	newquery := `BEGIN PACKAGE.StoredProc (:pnA1, :pnA2, :pnOutput); END;`

	stmt, err := db.PrepareContext(ctx, newquery)
	if err != nil {
		fmt.Printf("Prepare statement: %v\n", err)
		panic(err)
	}
	
	_, err = stmt.ExecContext(
		ctx,
		sql.Named("pnA1", pn1),
		sql.Named("pnA2", pn2),
		sql.Named("pnOutput", sql.Out{Dest: &outNo}),
	)

	if err != nil {
		fmt.Printf("statement execution: %v\n", err)
	}
	
	defer outNo.Close()

	cols1 := outNo.(driver.RowsColumnTypeScanType).Columns()
	fmt.Printf("%v", cols1)
	dests1 := make([]driver.Value, len(cols1))
	for {
		if err := outNo.Next(dests1); err != nil {
			if err == io.EOF {
				break
			}
			outNo.Close()
			fmt.Printf("Error: %v\n", err)
		}
		fmt.Println(dests1)
	}
}`

Output when running in Windows cmd is:
`statement execution: ORA-06550: line 1, column 7:
PLS-00306: wrong number or types of arguments in call to 'STOREDPROCADD'
ORA-06550: line 1, column 7:
PL/SQL: Statement ignored
panic: runtime error:...`

What am I missing?
Thanks.

**Comment by tgulacsi** [comment]
Oracle dows not return  CURSOR (*sql.Rows) in place of a NUMBER, so

```
newquery := `BEGIN PACKAGE.StoredProc (:pnA1, :pnA2, :pnOutput); END;`

_, err = stmt.ExecContext(
	ctx, newquery,
	sql.Named("pnA1", pn1),
	sql.Named("pnA2", pn2),
	sql.Named("pnOutput", sql.Out{Dest: &outNo}),
)

if err != nil {
	fmt.Printf("%s: %+v\n", newquery, err)
}

fmt.Println("out:", outNo)
```

should work.

**Reply by sindy-senorita** [reply]
Hi @tgulacsi . Thanks for sharing.
However, I got error using this.
This a snippet of what I do
(using gordor v.0.25.2)

```
import (
	"context"
	"database/sql"
	"github.com/godror/godror"
)


const (
	spQuery = `
DECLARE
	INUSERID VARCHAR2(12);
	OUTSOMEVALUE NUMBER;
BEGIN
DBMS_OUTPUT.enable(NULL);
SOME_SP(
	INUSERID => :INUSERID,
	OUTSOMEVALUE => :OUTSOMEVALUE
);
`
var outSomeValue float64
	_, err = conn.ExecContext(
		ctx,
		spQuery,
		sql.Named("INUSERID", userID),
		sql.Named("OUTSOMEVALUE", sql.Out{ Dest: &outSomeValue}),
	)
	if err != nil {
		return 0, err
	}
return outSomeValue, nil
```

this results in error
`sql: converting argument with name \"OUTSOMEVALUE\" type: unsupported type sql.Out, a struct`

Do you have any insight on this? Thanks in advance!

**Reply by tgulacsi** [reply]
I don't know. What if you leave out sql.Named, and use positional arguments?

**Reply by sindy-senorita** [reply]
@tgulacsi  it's the same
```
const (
	spQuery = `
DECLARE
	INUSERID VARCHAR2(12);
	OUTSOMEVALUE NUMBER;
BEGIN
SOME_SP(
	INUSERID => :1,
	OUTSOMEVALUE => :2
);
END;
`
var outSomeValue float64
	_, err = conn.ExecContext(
		ctx,
		spQuery,
		userID,
		sql.Out{ Dest: &outSomeValue},
	)
	if err != nil {
		return 0, err
	}
return outSomeValue, nil
```

throws `error="sql: converting argument $3 type: unsupported type sql.Out, a struct"`

However, I found that it is working if I prepare the statement (Just realized that your example is prepared). Both named and numbered param work ok with prepared statement
```
const (
	spQuery = `
DECLARE
	INUSERID VARCHAR2(12);
	OUTSOMEVALUE NUMBER;
BEGIN
SOME_SP(
	INUSERID => :1,
	OUTSOMEVALUE => :2
);
END;
`

conn, err := r.db.Conn(ctx)
if err != nil {
    return 0, err
}
defer func() { _ = conn.Close() }()
var outSomeValue float64
params := []interface{} {
	userID,
	sql.Out{ Dest: &outSomeValue},
}
stmt, err := conn.PrepareContext(ctx, spQuery)
if err != nil {
	return 0, err
}
defer stmt.Close()
_, err = stmt.ExecContext(
	ctx,
	params...,
)
if err != nil {
	return 0, err
}
return outSomeValue, nil
```

Is this expected? I'm curious why do prepared and unprepared work differently

**Reply by tgulacsi** [reply]
That "sql: converting argument" is from https://cs.opensource.google/go/go/+/refs/tags/go1.16.5:src/database/sql/convert.go;l=194
, but "unsupported type" has NOT FOUND under godror or database/sql.

What is your real database driver? Aren't you using some specific logging/ORM/tracing middleware driver?
Please try the code with vanilla "godror" driver only!

**Reply by sindy-senorita** [reply]
We connect with plain stdlib's sql: `sql.Open("godror", [DSN url])`, nothing in between...
Still curious.., but using prepared statements has fixed our immediate problem. Many thanks!

**Comment by sudarshan12s** [comment, marked as answer]
@kerocen, I have tried with linux, 19.3 instant client, It works without
prepare too. I haven't tried on windows though.

**Reply by sindy-senorita** [reply]
@sudarshan12s , mine is run on `golang:1.15` docker image, on top of mac OS
